### PR TITLE
Dev reports

### PR DESCRIPTION
**Pull request by ghost**
Creation of new pages for reports:
- list of tools + number of job in error + number of job in state "ok" per month ;
- list of tools + number of job in error + error messages ;
- list of tools + average execution time for the related jobs + last job's execution time + min and max execution time per month ;
- (list of histories + size + umber of dataset) for each user ;
- list of histories + number of datasets for each type (finished, paused, error, etc).

### CODE REVIEW

File: templates/webapps/reports/tools_and_job_state_peer_month.mako [hunk 1]
```diff
@@ -0,0 +1,44 @@
+<%inherit file="/base.mako"/>
+<%namespace file="/message.mako" import="render_msg" />
+
+%if message:
+    ${render_msg( message, 'done' )}
+%endif
+
+<div class="toolForm">
+    <h3 align="center">Number of Jobs Ok or in error peer month for ${tool}</h3>
```
Comment (nsoranzo) [inline templates/webapps/reports/tools_and_job_state_peer_month.mako:9]: s/peer/per/ 

Also in the file name.

File: lib/galaxy/webapps/reports/controllers/tools.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+        for date in dates:
+            date_key = date.strftime( "%B %Y" )
+            if date_key not in data:
+                data[date_key] = [int( date_and_jobs_ok.get (date, 0) ), int( date_and_jobs_error.get (date, 0) ) ]
+            else :
+                data[date_key][0] += int( date_and_jobs_ok.get ( date, 0 ) )
+                data[date_key][1] += int( date_and_jobs_error.get( date, 0 ) )
+        
+
+        return trans.fill_template( '/webapps/reports/tools_and_job_state_peer_month.mako',
```
Comment (nsoranzo) [inline lib/galaxy/webapps/reports/controllers/tools.py:182]: s/peer/per/

File: lib/galaxy/webapps/reports/controllers/tools.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+                "last": self.formated (data[tool]["last"], color)}
+
+        return trans.fill_template( '/webapps/reports/tool_execution_time.mako',
+                                    data=ordered_data,
+                                    descending=descending,
+                                    user_cutoff=user_cutoff,
+                                    sort_by=sort_by )
+
+    @web.expose
+    def tool_execution_time_peer_month (self, trans, **kwd):
```
Comment (nsoranzo) [inline lib/galaxy/webapps/reports/controllers/tools.py:253]: s/peer/per/

File: lib/galaxy/webapps/reports/controllers/tools.py [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+        """
+        Fill th template tool_execution_time_peer_month.mako with informations:
```
Comment (nsoranzo) [inline lib/galaxy/webapps/reports/controllers/tools.py:255]: s/th/the/
s/peer/per/

File: lib/galaxy/webapps/reports/controllers/tools.py [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+        months.sort (sort_functions[sort_by])
+        if user_cutoff:
+            months = months[:user_cutoff]
+
+        for month in months:
+            ordered_data[str(month[0]).split(' ')[0][:-3]] = (self.formated (month[1], color),
+                self.formated (month[2], color),
+                self.formated (month[3], color))
+
+        return trans.fill_template( '/webapps/reports/tool_execution_time_peer_month.mako',
```
Comment (nsoranzo) [inline lib/galaxy/webapps/reports/controllers/tools.py:292]: s/peer/per/

File: templates/webapps/reports/tool_execution_time.mako [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            </tr>
+            <% odd = False%>
+            %for tool in data:
+                %if odd:
+                    <tr class="odd_row">
+                %else:
+                    <tr class="tr">
+                %endif
+                <td>
+                <a href=tool_execution_time_peer_month?tool=${tool}>${tool}</a>
```
Comment (nsoranzo) [inline templates/webapps/reports/tool_execution_time.mako:67]: s/peer/per/

File: templates/webapps/reports/tool_execution_time_peer_month.mako [hunk 7]
```diff
@@ -0,0 +1,75 @@
+<%inherit file="/base.mako"/>
+<%namespace file="/message.mako" import="render_msg" />
+
+
+<div class="toolForm">
+    <h3 align="center">Time execution for ${tool} peer month</h3>
```
Comment (nsoranzo) [inline templates/webapps/reports/tool_execution_time_peer_month.mako:6]: s/peer/per/

Also in the file name.

File: templates/webapps/reports/tool_execution_time_peer_month.mako [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    %elif sort_by == 2:
+        max time
+    %elif sort_by == 3:
+        average time
+    %endif
+    </h4>
+    <table align="center" width="70%" class="colored" cellpadding="5" cellspacing="5">
+        <tr>
+            <td>
+                <form method="post" controller="users" action="tool_execution_time_peer_month?tool=${tool}">
```
Comment (nsoranzo) [inline templates/webapps/reports/tool_execution_time_peer_month.mako:27]: s/peer/per/

File: templates/webapps/reports/tools_and_job_state.mako [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            <% odd = False%>
+            %for tool in data:
+                %if odd:
+                    <tr class="odd_row">
+                %else:
+                    <tr class="tr">
+                %endif
+                <td>
+                %if data[tool][0] + data[tool][1] != "--":
+                    <a href=tools_and_job_state_peer_month?tool=${tool}>${tool}</a>
```
Comment (nsoranzo) [inline templates/webapps/reports/tools_and_job_state.mako:65]: s/peer/per/

File: lib/galaxy/webapps/reports/controllers/tools.py [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+                    return '<font color="orange">' + returned + '</font>'
+                return returned
+            if splited[1]:
+                return "%d min. %d sec." % splited[1:3]
+            return "%.1f sec." % splited[2]
+
+    @web.expose
+    def tools_and_job_state( self, trans, **kwd ):
+        """
+        fill tools_and_job_state_peer_month.mako template with
```
Comment (nsoranzo) [inline lib/galaxy/webapps/reports/controllers/tools.py]: s/peer/per/

Comment (nsoranzo) [inline lib/galaxy/webapps/reports/controllers/tools.py:69]: s/tools_and_job_state_peer_month/tools_and_job_state/

File: lib/galaxy/webapps/reports/controllers/tools.py [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+
+        return trans.fill_template( '/webapps/reports/tools_and_job_state.mako',
+                                    data=data,
+                                    user_cutoff=user_cutoff,
+                                    sorting=sorting,
+                                    descending=descending,
+                                    message=message )
+
+    @web.expose
+    def tools_and_job_state_peer_month (self, trans, **kwd ):
```
Comment (nsoranzo) [inline lib/galaxy/webapps/reports/controllers/tools.py:123]: s/peer/per/

File: lib/galaxy/webapps/reports/controllers/tools.py [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+        """
+        fill tools_and_job_state_peer_month.mako template with
```
Comment (nsoranzo) [inline lib/galaxy/webapps/reports/controllers/tools.py:125]: s/peer/per/

**Comment by dannon** [comment]
Thanks for your interest in the reports app @BalthazarPavot.  I'm writing a quick PR that'll bring the history.py up to par to pass lint and will open it in a sec.  Hopefully from there, you can take it and bring the other files up to par as well.  The easiest way to check this prior to submission is to just use 'flake8' locally on the file.

**Comment by dannon** [comment]
Ok, see https://github.com/BalthazarPavot/galaxy/pull/1 for my changes to history.py -- I'll leave tools.py up to you!

File: lib/galaxy/webapps/reports/controllers/history.py [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+import calendar, operator
+import sqlalchemy as sa
+import logging
+from datetime import datetime, date, timedelta
+from time import strftime
+
+import galaxy.model
+import pkg_resources
+from galaxy import util
+from galaxy.model.orm import and_, not_, or_
```
Comment (martenson) [inline lib/galaxy/webapps/reports/controllers/history.py]: this should be `from sqlalchemy import and_`

File: lib/galaxy/webapps/reports/controllers/tools.py [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+import sqlalchemy as sa
+import logging
+from datetime import datetime, date, timedelta
+from time import strftime
+import collections
+
+import galaxy.model
+import pkg_resources
+from galaxy import util
+from galaxy.model.orm import and_, not_, or_
```
Comment (martenson) [inline lib/galaxy/webapps/reports/controllers/tools.py:13]: this also should be `from sqlalchemy import and_`

File: templates/webapps/reports/history_and_dataset_per_user.mako [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+                            <option value="size"> by history space </option>
+                            <option value="HSort"> by number of history </option>
+                            <option value="DSort"> by number of dataset </option>
+                        </select>
+                        <select value="${descending}" size="3" name="descending">
+                            <option value="desc"> descending </option>
+                            <option value="asc"> ascending </option>
+                        </select>
+                        </br>
+                        <button name="action" value="commit">Sort my Data!</button>
```
Comment (martenson) [inline templates/webapps/reports/history_and_dataset_per_user.mako:46]: please remove the `!` and optionally drop the `my Data` too

File: templates/webapps/reports/index.mako [hunk 16]
```diff
@@ -74,6 +74,28 @@
...
+                    </div>
+                </div>
+                <div class="toolSectionPad"></div>
+                <div class="toolSectionPad"></div>
+                <div class="toolSectionTitle">
+                    <span>Tools</span>
+                </div>
+                <div class="toolSectionBody">
+                    <div class="toolSectionBg">
+                        <div class="toolTitle"><a target="galaxy_main" href="${h.url_for( controller='history', action='history_and_dataset_per_user' )}">History and dataset per user</a></div>
```
Comment (martenson) [inline templates/webapps/reports/index.mako:92]: this is twice here, once under History and once under Tools, same with the menu option below

**Comment by martenson** [comment]
Nice work @BalthazarPavot - please see the comments inline. 
please note that ultimately this needs to target the `dev` branch

preview
screenshot 2016-05-10 10 22 08

**Comment by jmchilton** [comment]
Thanks for the contribution - I'm attaching a WIP tag until the comments above are addressed but this PR will need to be closed and a new one opened to target `dev` as @martenson mentioned.

**Comment by martenson** [comment]
closing in favor of https://github.com/galaxyproject/galaxy/pull/2592

**Pull request closed without merging**